// @dynamic-labs/sdk-react-core/src/lib/utils/hooks/useSendDynamicProps/useSendDynamicProps.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
require('@dynamic-labs/sdk-api');
var logger = require('../../../shared/logger.cjs');
require('@dynamic-labs/iconic');
require('@dynamic-labs/wallet-connector-core');
require('react/jsx-runtime');
require('../../../context/ViewContext/ViewContext.cjs');
require('@dynamic-labs/wallet-book');
require('../../constants/colors.cjs');
require('../../constants/values.cjs');
require('../../../shared/utils/classes/storage/localStorage.cjs');
var useEffectOnce = require('../../../shared/utils/hooks/useEffectOnce/useEffectOnce.cjs');
require('@dynamic-labs/utils');
require('../../../shared/consts/index.cjs');
var api = require('../../../data/api.cjs');

const useSendDynamicProps = ({ settings, environmentId, }) => {
    useEffectOnce.useEffectOnce(() => {
        // we ReactNode components are not being serialized properly
        // remove them for now and deal with them later
        // context: https://linear.app/dynamic-labs/issue/GVTY-328/[starknet]-sdk-api-is-not-serializing-react-components
        // future work: https://linear.app/dynamic-labs/issue/GVTY-330/bug-figure-out-why-react-components-are-not
        const updatedDynamicProps = {
            settings: Object.assign(Object.assign({}, settings), { customPrivacyPolicy: Boolean(settings.customPrivacyPolicy), customTermsOfServices: Boolean(settings.customTermsOfServices), policiesConsentInnerComponent: Boolean(settings.policiesConsentInnerComponent) }),
        };
        const serializedEventsCallbacks = serializeEventsCallbacks(settings.eventsCallbacks);
        const serializedWalletConnectors = serializeWalletConnectors(settings.walletConnectors);
        const serializedDynamicContextProps = Object.assign(Object.assign({}, updatedDynamicProps), { settings: Object.assign(Object.assign({}, updatedDynamicProps.settings), { eventsCallbacks: serializedEventsCallbacks, walletConnectors: serializedWalletConnectors }) });
        api.sendDynamicProps(environmentId, {
            dynamicContextProps: serializedDynamicContextProps,
            frameworkSettings: getFrameworkSettings(),
        });
    });
};
const serializeEventsCallbacks = (eventsCallbacks) => {
    try {
        if (!eventsCallbacks)
            return undefined;
        const eventsCallbackValues = Object.values(eventsCallbacks);
        const eventsCallbacksKeys = Object.keys(eventsCallbacks);
        return eventsCallbacksKeys.filter((key, index) => {
            if (eventsCallbackValues[index] === undefined) {
                return false;
            }
            return key;
        });
    }
    catch (error) {
        logger.logger.error('error serializing eventsCallbacks', error);
        return undefined;
    }
};
const serializeWalletConnectors = (walletConnectors) => {
    try {
        if (!walletConnectors)
            return undefined;
        return walletConnectors.map((connector) => connector.name);
    }
    catch (error) {
        logger.logger.error('error serializing walletConnectors', error);
        return undefined;
    }
};
const getFrameworkSettings = () => {
    var _a;
    return ({
        nextJs: {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            version: (_a = window.next) === null || _a === void 0 ? void 0 : _a.version,
        },
        react: {
            version: React.version,
        },
    });
};

exports.serializeEventsCallbacks = serializeEventsCallbacks;
exports.serializeWalletConnectors = serializeWalletConnectors;
exports.useSendDynamicProps = useSendDynamicProps;
